---
title: "IsoDeconvMM Pipeline"
author: "Hillary Heiling"
date: "October 16, 2019"
output: html_document
---

## Items needed before Step 0:

* .bam files of mixture and pure samples (not .bam.bai)
* Cuffdiff output

## Original Process: Before GeneModel_Build.R file and corresponding processes

*Step 0.1*: step1_check_and_filter.R



*Step 0.2*: step2_read_depth.R

*Step 0.3*: step3_countReads.R

*Step 0.4*: step4_summarize_counts.r

*Step 0.5*: FragLengths_Function_mixutre.R

## Updated Process: Step 0

Step 0.1:

```{r}

```


## Original Process: GeneModel_Build and later

*Step 1*:  Mixture_Creation / GeneModel_Build.R 

Calls dev_compiles_geneMod() function from isoDeconv_geneModel_revised.R file

Inputs: 

Output: fit_geneMod object

Note: changed output of dev_compiles_geneMod() to be just an object, don't save an .RData object to some folder

*Step 2*:  Mixture_Creation / Restricting_Geneids_Pairwise.R

Establishes transcript clusters with highest levels of discriminatory capability

Library needed: biocLite.R library cummerbund

Inputs: "list_of_genes.txt" name annotation file, Cuffdiff_Out folder

Output: finalOut2 object as EnsembleIds2Use.RData object

*Step 3*:  Process_Real_Data / prepare_sigred.R

Inputs:  fin_geneMod object from *Step 1*, finalOut2 (EnsembleIds2Use.RData) object from *Step 2*

Output: sig_geneMod

*Step 4*:  Mixture_Creation / RecharacterizedOutput.R

Input: The sig_geneMod object from Step 3

Output:  Slightly modified version of sig_geneMod object

*Step 5*: Run pure sample production functions

Input: The modified sig_geneMod object from Step 4

Ouput: tmp.data = output from Pure.apply.fun() function from pure sample production functions file.

## Updated Process: GeneModel_Build.R and later

Step 1:

Note: need to generalize to multiple files (multiple combinations of one mixture and all cell type references)

```{r}
# Required Libraries:
library(gtools)
library(IsoDeconvMM)

# Inputs
file = "g10h90" # Should be one of the comboLabels provided in FragLengths_Function_mixture.R

sys_statement1 = sprintf("countData=c(\"mf_%s_counts.txt\",
                         \"CD4_SRR1550995_SBC_counts.txt\",
                         \"CD4_SRR1551016_SBC_counts.txt\",
                         \"CD4_SRR1551098_SBC_counts.txt\",
                         \"CD8_SRR1551024_SBC_counts.txt\",
                         \"CD8_SRR1551051_SBC_counts.txt\",
                         \"CD8_SRR1551065_SBC_counts.txt\")",file)
eval(parse(text=sys_statement1))
labels = c("mix","CD4_ref1","CD4_ref2","CD4_ref3",
           "CD8_ref1","CD8_ref2","CD8_ref3")
cellTypes = c("mix","CD4","CD4","CD4",
              "CD8","CD8","CD8")
fragSizeFile = sprintf("/netscr/drwilson/2018-04-05 Paper 1/Frag_Lengths/%s_fraglens.txt",file)
bedFile = "/netscr/drwilson/Reference_Annotations/Homo_sapiens/Homo_sapiens.GRCh37.66.nonoverlap.exon.bed"
knownIsoforms = "/netscr/drwilson/Reference_Annotations/Homo_sapiens/Homo_sapiens.GRCh37.66.nonoverlap.exon.knownIsoforms.RData"
readLen=50
lmax=600
eLenMin=1

```

Suppose the above process has been generalized to multiple files

```{r}
files = file

final_geneMod = list()

for(j in 1:length(files)){
  # Call dev_compiled_geneMod function
  fin_geneMod = dev_compiled_geneMod(countData=countData,labels = labels,total_cts = total_cts, 
                       cellTypes=cellTypes, bedFile=bedFile,knownIsoforms=knownIsoforms,
                       fragSizeFile=fragSizeFile,output=output,readLen=readLen,lmax=lmax,
                       eLenMin=eLenMin)
  final_geneMod[[j]] = fin_geneMod
}

```


Step 2:

```{r}
#-----------------------------------------------------------------------------#
# ESTABLISHING CLUSTERS WITH HIGHEST LEVELS OF DISCRIMINATORY CAPABILITY      #
#-----------------------------------------------------------------------------#
#------------------ LOAD AND PROCESS GENE NAMES -------------------#
anno_names = read.table(file = "D:/DougData/Documents/Genomics Training Grant/Update_12_10_2014/FINAL/Article/list_of_genes.txt",header = TRUE,sep = ",")
anno_names$gene_short_name = anno_names$Associated.Gene.Name

#------------------ CALL THE LIBRARY ------------------#
source("https://bioconductor.org/biocLite.R")
biocLite("cummeRbund")
library(cummeRbund)

# Had to downgrade to RSQ-lite v 1.1-2

finalOut2 = EnsemblIds2Use(folder = "Cuffdiff_Out",
                            directory = "D:/DougData/Documents/Dissertation/Paper 1 - Cell Type Abundance Estimation/Daily Work/1_30_2018/")

```

Step 3:

```{r}
setwd("/netscr/drwilson/2018-04-05 Paper 1/MappedData/Real_Data_Opt/GeneModels/")

analy_genes = finalOut2$Ensembl.ID

significant_geneMod = list()

for(j in 1:length(final_geneMod)){

  fin_geneMod = final_geneMod[[j]]
  
  indices2chk = which(names(fin_geneMod)!="Sample_Info")
  indices_tmp = NULL
  indices=NULL
  indices_tmp = rep(0,length(geneMod)) # What is this geneMod object?
  for(i in indices2chk){
    infodf = fin_geneMod[[i]]$info
    genesi = unique(infodf$gene)
    genesi = unique(unlist(strsplit(x=genesi,split = ":")))
    if(any(genesi %in% analy_genes)){indices_tmp[i]=1}
  }
  indices = which(indices_tmp==1)
  
  sig_geneMod = fin_geneMod[indices]
  sig_geneMod["Sample_Info"] = fin_geneMod["Sample_Info"]
  
  sig_geneMod = rem_clust(geneMod = sig_geneMod,co = 5,min_ind = 0)
  
  # message("File: ",file_labels[j]," // nGenes= ",length(sig_geneMod))
  # 
  # save(sig_geneMod,file=file_labels[j])
  
  significant_geneMod[[j]] = sig_geneMod
}
```

Step 4:

```{r}

library(gtools)

#-------------------------------------------------------------------#
# EDIT TO GROUP CELL TYPES                                          #
#-------------------------------------------------------------------#

modified_sig_geneMod = list()

for(f in 1:length(significant_geneMod)){
  
  sig_geneMod = significant_geneMod[[f]]
  
  info_mat = sig_geneMod[["Sample_Info"]]
  cellTypes = unique(info_mat$Cell_Type)
  
  ctList = list()
  
  for(j in 1:length(cellTypes)){
    idx = which(info_mat$Cell_Type==cellTypes[j])
    ctList[[cellTypes[j]]] = list(samps = info_mat$Label[idx], tots = info_mat$Total[idx])
  }
  
  idx2consider = which(names(sig_geneMod)!="Sample_Info")
  for(k in idx2consider){
    for(l in 1:length(cellTypes)){
      samps2use = ctList[[l]]$samps
      tots      = ctList[[l]]$tots
  
      y_vecs  = paste("sig_geneMod[[k]]$y",samps2use,sep = "_")
      y_vecsc = paste(y_vecs,collapse = ",")
      nExon = eval(parse(text=sprintf("length(%s)",y_vecs[1])))
      textcmd = sprintf("matrix(c(%s),nrow=nExon,ncol=length(samps2use))",y_vecsc)
      expMat  = eval(parse(text=textcmd))
  
      totmg   = tots-colSums(expMat)
      expMat2 = rbind(totmg,expMat)
  
      if(cellTypes[l]!="mix"){
        sig_geneMod[[k]][[cellTypes[l]]] = list(cellType=cellTypes[l],rds_exons=expMat2)
      } else {
        sig_geneMod[[k]][[cellTypes[l]]] = list(cellType=cellTypes[l],rds_exons_t=expMat2)
      }
      
    }
  }
  
  modified_sig_geneMod[[f]] = sig_geneMod
  # save(sig_geneMod,file=sprintf("rechar_%s_geneMod.RData",file))
}

```

Step 5:

Note: Confused about what files are used here

```{r}
# Call_Model / Revised_Sim_PS_codeNew.R code process

library(gtools)
library(alabama)

#-----------------------------------------------------------#
# CALL Pure Sample                                          #
#-----------------------------------------------------------#
cellTypes = c("cd4","cd8")

pure_est = list()

for(j in 1:length(modified_sig_geneMod)){
  
  sig_geneMod = modified_sig_geneMod[[j]]
  
  sim.out = sig_geneMod[which(names(sig_geneMod)!="Sample_Info")]

  # Clusters with single isoforms:
  # EXCLUDE THEM FOR THE MOMENT!.
  dim_mat = matrix(0,nrow=length(sim.out),ncol=2)
  excl_clust = c()
  excl_clust2 = c()
  for(i in 1:length(sim.out)){
    dim_mat[i,] = dim(sim.out[[i]][["X"]])
    if(all(dim_mat[i,]==c(1,1))){
      excl_clust = c(excl_clust,i)
    }
    if(dim_mat[i,2] == 1){
      excl_clust2 = c(excl_clust2,i)
    }
  }
  
  excl_clust_union = union(excl_clust,excl_clust2)
  if(length(excl_clust_union)>0){
    sim.new = sim.out[-c(excl_clust_union)]
  } else {
    sim.new = sim.out
  }
  
  
  # Optimize the Pure Sample Functions:
  tmp.data = Pure.apply.fun(data.list = sim.new, cellTypes = cellTypes, corr_co = 1)
  
  # save_cmd = sprintf("save(tmp.data,file=\"../Compiled_Output/%s_pure_est.RData\")",file)
  # eval(parse(text=save_cmd))
  
  pure_est[[j]] = tmp.data
  
  # message("Sample ",file,"Completed")
    
}



```

Step 6:

Multiple Cluster code: (Revised_Sim_Mix_Calls_MI.R)

Calls Single Cluster code: (Revised_Sim_MixCode_SI.R)

```{r}

library(alabama)

IsoDeconv_Output = list()

for(i in 1:length(pure_est)){
  
  tmp.data = pure_est[[i]]
  
  #--------------------------------------------------------#
  # Establish input break ups                              #
  #--------------------------------------------------------#
  # Cell-Types:
  cellTypes = c("cd4","cd8")
  
  # Data Set Necessities:
  clust.start = 1
  clust.end = length(tmp.data)
  by.value = 15
  
  start.pts = seq(from = 1,to = clust.end,by = by.value)
  end.pts = c((start.pts[-1]-1),clust.end)
  
  cluster_output = list()
  for(m in 1:length(start.pts)){
    
    # Call Revised_Sim_MixCode_SI.R code
    curr.clust.opt = tmp.data[c(start.pt:end.pt)]
    curr.clust.out = STG.Update_Cluster.All(all_data=curr.clust.opt, cellTypes = cellTypes,
                                            optimType="nlminb", simple.Init=FALSE, initPts=c(0.5))
    
    cluster_output[[m]] = curr.clust.out
  }
  
  IsoDeconv_Output[[i]] = cluster_output
}

```


## Edit of Updated Process

Note: Once I'm confident the "Updated Process" steps above work as expected, I will modularize some steps (create some additional functions). Ideally, it would be nice to have just one "wrapper" function where all of the inputs go and one output.

The End
